harvester_e2e_tests/integration/vmconsole.py

```diff
@@ -0,0 +1,60 @@
+import re
+from time import sleep
+from datetime import datetime, timedelta
+from subprocess import Popen, PIPE
+
+
+class VMConsole:
+    def __init__(self, virctl, name, user, passwd, command_timeout=300):
+        self.proc = None
+        self.virctl = virctl
+        self.name = name
+        self.user = user
+        self.passwd = passwd
+        self.timeout = command_timeout
+
+    def __repr__(self):
+        return (f"{__class__.__name__}({self.virctl!r}, {self.name!r}"
+                f", {self.user!r}, {self.passwd!r}, {self.timeout})")
+
+    def __enter__(self):
+        self.login()
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        self.logout()
+
+    def login(self, timeout=None, **kwargs):
+        self.proc = Popen(f"{self.virctl} console {self.name} 2>&1",
+                          stdin=PIPE, stdout=PIPE, shell=True, **kwargs)
+
+        endtime = datetime.now() + timedelta(seconds=timeout or self.timeout)
+        while endtime > datetime.now():
+            out = self.execute_command("\n\n")
+            if re.search(rf"{self.name} login:\s+?", out, re.M):
+                break
+            print(out)
+            sleep(5)
+        else:
+            raise TimeoutError(-1, "Login timeout: Unable to catch login hints.\n"
+                               f"self.{out}")
+
+        out = self.execute_command(self.user)
+
+        return self.execute_command(self.passwd)
+
+    def logout(self):
+        if self.proc:
+            self.execute_command("exit\n")
+            self.proc.communicate()
+            self.proc = None
+
+    def execute_command(self, command, *, timeout=None):
+        if not self.proc:
+            self.login(timeout)
+        proc = self.proc
+        proc.stdin.write(command.encode() + b"\n")
+        proc.stdin.flush()
+        sleep(1)
+        outbytes = len(proc.stdout.peek())
+        return proc.stdout.read(outbytes).decode()
```

Comment:
This is awesome!!!
I really like the idea that we have here with leveraging `virtctl` to interact with the VMs on Harvester given that we have the respective `kubeconfig` and other credentials for the VM.
Really excellent find @lanfon72 ! :+1: :raised_hands: 

This reduces so much complexity and overhead of needing to deal with waiting for the VM to be up, grabbing the IPv4 associated with the VM and then juggling something like a Paramiko connection out to it.  Especially if we're doing something fairly complex with the VM post cloud-init stage.  Since we're leveraging Kubevirt behind the scenes on Harvester, `virtctl` as a binary that gets leveraged in this context-manager wrapper is just awesome!

Like if we wanted to do something more complex with actually writing a ton of data to the VM's disk directly once it's up prior to migrating, say like writing 5GB of data so the disk is at least somewhat full:
```
    cpu, mem = 1, 2
    vm = api_client.vms.Spec(cpu, mem)
    vm.add_image("disk-0", image_id)

    code, data = api_client.vms.create(unique_vm_name, vm)

    assert 201 == code, (code, data)

    endtime = datetime.now() + timedelta(seconds=wait_timeout)
    while endtime > datetime.now():
        code, data = api_client.vms.get_status(unique_vm_name)
        if 200 == code and "Running" == data.get('status', {}).get('phase'):
            # some logic to check whether we have network to get virtctl or we need to use a file, if we have network
            # maybe we grabbed it earlier during the testing phase and set it at a global level 
            # as to not try to regrab it again - maybe it's just globally accesible or a fixture?
            with VMConsole(VIRTCTL_GLOBAL, unique_vm_name, ubuntu, ubuntupw, 600) as unique_vm_console:
                write_file_result = unique_vm_console.execute_command('dd if=/dev/urandom of=/home/ubuntu/sample.txt bs=5G count=1')
                # and then other checks or something...
        sleep(3)
    else:
        raise AssertionError(
            f"Failed to create Minimal VM({cpu} core, {mem} RAM) with errors:\n"
            f"Phase: {data.get('status', {}).get('phase')}\t"
            f"Status: {data.get('status')}\n"
            f"API Status({code}): {data}"
        )
```

`execute_command` will really help us to do more complicated testing with VMs on Harvester.
If we want to later, we might be able to expand the `sleep(1)` to be more of an exponential backoff - or just bring in a utility package like [backoff](https://github.com/litl/backoff) and add the decorator to the class function/method - if we're executing a command that may be like "long running" or something like fetching a lot of content via `wget` or something to avoid being rate-limited.

Reply:
Commands executing through the subprocess, backoff will not work.

The Console is not much stable to execute command, so I didn't use it to manipulate VMs.
(But the implementation will still help us to monitor VM's information while rebooting)

Reply:
We could look into finding a way to make the console more stable later on :smile:. 
I agree that the implementation will help us monitor VM's info during rebooting.
If we could later on use it to manipulate VMs more, that could help expand the depth of our tests, like with the example I was sorta thinking about surrounding us being able to actually write a file to the disk via the console execute_command, vs needing to juggle SSH connection out through a separate library like Paramiko or something.